### ISSUE TITLE
How hook a forwarded connection closing

### ISSUE THREAD

**Issue by jabbera** [ronf/asyncssh#557]
Hi! I was hoping you could help me out. I'm using asyncssh as a jump server. I have users starting to use a single jump connection to forward to multiple targets. I do some state management around those targets, such as making sure they are up before forwarding the connection and notifying an orchestrator that ssh sessions are still in progress.

Now that users are multiplexing over a single jump connection as opposed to using a jump connection for each target I'm trying to understand the best way to understand when a forwarded connection is shut down: IE

ssh -J my_custom_broker_program target

How do I know when the connection to the target is terminated even though the jump connection at my_custom_broker_program is up.

Pseudo code below. I'm basically looking for a connection_lost event on the SSHFowarder returned by forward_connection.

```
class MySSHServer(asyncssh.SSHServer).

  async def __create_server_and_forward(self, dest_host: str, dest_port: int) -> asyncssh.SSHForwarder:
    return await self.__conn.forward_connection(dest_host, dest_port)

  def connection_requested(
        self, dest_host: str, dest_port: int, orig_host: str, orig_port: int
  ) -> Coroutine[Any, Any, asyncssh.SSHForwarder] | bool:
      return self.__create_server_and_forward(dest_host, dest_port)
```

**Comment by jabbera** [comment]
I went back in a previous issue we worked together and found this: https://github.com/ronf/asyncssh/issues/471#issuecomment-1086882021

however it seems wrong since I need to access a private variable.

**Comment by ronf** [comment]
You can always get the current event loop from inside any async function by calling `asyncio.get_event_loop()`, so there's no need to access the connection's private member for that. However, there's still the issue I mentioned around half-open connections to deal with.

More generally, though, how would you get any notification from this forwarder object? If you are using the return value from `conn.forward_connection()` directly, it's going to return an AsyncSSH `SSHForwarder` which is meant to be self-contained. It doesn't have the concept of an "owner" object to send any notifications to. It will do things like pass on both EOF and close messages to its peer for you, so everything should be getting cleaned up properly, but if you want a "side" notification that this is happening, you'd need you have your own subclass and right now there's no provision for that in the `conn.forward_connection()` method. You'd have to something like monkey-patch the class.

Are you trying to determine when to shut down the upstream SSH connection that you opened, doing so when the last forwarded connection to that upstream server closes?

There isn't much to the `forward_connection()` method, so creating your own version of that which has its own SSHForwarder subclass wouldn't be too bad, if you wanted to override some of the methods there. You could then track when `close()` is called on that object. It's better to use that in this case rather than `connection_lost()`, as close can be triggered from the peer as well as from `connection_lost()`. When all the upstream tunnels are closed, you could close the corresponding SSH connection.

**Comment by jabbera** [comment]
> Are you trying to determine when to shut down the upstream SSH connection that you opened, doing so when the last forwarded connection to that upstream server closes?

Nope. I'm using asyncssh as a jump host, and when someone creates a new tunnel I ensure the backend compute for that target is spun up and ready to go before the forward request is fulfilled. IE: ssh -J asyncsshjumphost backend_compute. I'd make sure backend_compute is spun up and ready to go. Each forwarded connection has status and metrics I report on at a regular interval that's done from asyncssh, and when all connections to a single target are closed I shutdown the dyamically allocated compute and save it's state.

My initial implementation only supported 1 tunnel per peer at a time. I'm supporting multiple now hence my need to understand the lifetime of the tunnel connection. I ended up subclassing SSHForwarder and doing what you suggested, but I've overridden connection_lost as opposed to closed, thanks for the guidance here!

**Comment by ronf** [comment]
Yeah - try overriding the `close()` method on the SSHForwarder subclass and see if that gives you what you need.

**Comment by jabbera** [comment]
> Yeah - try overriding the `close()` method on the SSHForwarder subclass and see if that gives you what you need.

Worked great! Thanks again.
